java-checks/src/main/java/org/sonar/java/checks/CollectionMethodsWithLinearComplexityCheck.java

```diff
@@ -0,0 +1,142 @@
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.VariableTree;
+
+import javax.annotation.CheckForNull;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+@Rule(key = "S2250")
+public class CollectionMethodsWithLinearComplexityCheck extends IssuableSubscriptionVisitor {
+
+  private static final String JAVA_LANG_OBJECT = "java.lang.Object";
+  private static final String ARRAY_LIST = "java.util.ArrayList";
+  private static final String LINKED_LIST = "java.util.LinkedList";
+  private static final String COW_ARRAY_LIST = "java.util.concurrent.CopyOnWriteArrayList";
+  private static final String COW_ARRAY_SET = "java.util.concurrent.CopyOnWriteArraySet";
+  private static final String CONCURRENT_LINKED_QUEUE = "java.util.concurrent.ConcurrentLinkedQueue";
+  private static final String CONCURRENT_LINKED_DEQUE = "java.util.concurrent.ConcurrentLinkedDeque";
+
+  private static MethodMatcher collectionMethodMatcher() {
+    return MethodMatcher.create().typeDefinition(TypeCriteria.subtypeOf("java.util.Collection"));
+  }
+
+  private static final Map<MethodMatcher, Set<String>> matcherActualTypeMap;
+  static {
+    ImmutableMap.Builder<MethodMatcher, Set<String>> builder = ImmutableMap.builder();
+
+    MethodMatcher collectionContains = collectionMethodMatcher().name("contains").addParameter(JAVA_LANG_OBJECT);
+    builder.put(collectionContains, ImmutableSet.of(ARRAY_LIST, LINKED_LIST, COW_ARRAY_LIST, COW_ARRAY_SET, CONCURRENT_LINKED_QUEUE, CONCURRENT_LINKED_DEQUE));
+
+    MethodMatcher collectionSize = collectionMethodMatcher().name("size").withoutParameter();
+    builder.put(collectionSize, ImmutableSet.of(CONCURRENT_LINKED_QUEUE, CONCURRENT_LINKED_DEQUE));
+
+    MethodMatcher collectionAdd = collectionMethodMatcher().name("add").addParameter(TypeCriteria.anyType());
+    builder.put(collectionAdd, ImmutableSet.of(COW_ARRAY_SET, COW_ARRAY_LIST));
+
+    MethodMatcher collectionRemove = collectionMethodMatcher().name("remove").addParameter(JAVA_LANG_OBJECT);
+    builder.put(collectionRemove, ImmutableSet.of(ARRAY_LIST, COW_ARRAY_SET, COW_ARRAY_LIST));
+
+    MethodMatcher listGet = MethodMatcher.create().typeDefinition(TypeCriteria.subtypeOf("java.util.List")).name("get").addParameter("int");
+    builder.put(listGet, ImmutableSet.of(LINKED_LIST));
+    matcherActualTypeMap = builder.build();
+  }
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return Collections.singletonList(Tree.Kind.METHOD_INVOCATION);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    MethodInvocationTree mit = (MethodInvocationTree) tree;
+    if (!invocationInMethod(mit)) {
```

Comment:
Maybe add a comment to make explicit that you do not consider invocation in constructors, initializers and static initializers.

Now, I would first check if the method is matching any of the matcher before checking if the invocation is part of a method. Intuitively, I have the feeling that we will then do that computation less time, as the probability that a method invocation matches any of the methods is way less than the the probability of being part of a constructor.